File: packages/cc-headless/src/alarm-parser.ts
import type { AlarmContext } from './prompt-builder.js';

interface CloudWatchAlarmSns {
  AlarmName?: string;
  NewStateValue?: string;
  NewStateReason?: string;
  StateChangeTime?: string;
  Region?: string;
  Trigger?: {
    MetricName?: string;
    Namespace?: string;
    Dimensions?: Array<{ name: string; value: string }>;
    Statistic?: string;
    Period?: number;
    Threshold?: number;
    ComparisonOperator?: string;
  };
}

function toDimensions(
  dims?: Array<{ name: string; value: string }>,
): Record<string, string> | undefined {
  if (!dims || dims.length === 0) return undefined;
  return Object.fromEntries(dims.map((d) => [d.name, d.value]));
}

export function parseAlarm(data: Record<string, unknown>): AlarmContext {
  const alarm = data as unknown as CloudWatchAlarmSns;

  return {
    alarmName: alarm.AlarmName ?? 'UnknownAlarm',
    stateReason: alarm.NewStateReason ?? '',
    stateChangeTime: alarm.StateChangeTime,
    region: alarm.Region ?? process.env.AWS_REGION ?? 'us-east-1',
    metricName: alarm.Trigger?.MetricName,
    namespace: alarm.Trigger?.Namespace,
    dimensions: toDimensions(alarm.Trigger?.Dimensions),
    statistic: alarm.Trigger?.Statistic,
    period: alarm.Trigger?.Period,
    threshold: alarm.Trigger?.Threshold,
    comparisonOperator: alarm.Trigger?.ComparisonOperator,
  };
}

File: packages/cc-headless/src/cc_headless/alarm_parser.py
from __future__ import annotations

import os
from dataclasses import dataclass, field


@dataclass
class AlarmContext:
    alarm_name: str = "UnknownAlarm"
    state_reason: str = ""
    state_change_time: str | None = None
    region: str = "us-east-1"
    metric_name: str | None = None
    namespace: str | None = None
    dimensions: dict[str, str] = field(default_factory=dict)
    statistic: str | None = None
    period: int | None = None
    threshold: float | None = None
    comparison_operator: str | None = None


def parse_alarm(data: dict) -> AlarmContext:
    trigger = data.get("Trigger", {}) or {}
    dims_raw = trigger.get("Dimensions") or []
    dimensions = {d["name"]: d["value"] for d in dims_raw if "name" in d and "value" in d}

    return AlarmContext(
        alarm_name=data.get("AlarmName", "UnknownAlarm"),
        state_reason=data.get("NewStateReason", ""),
        state_change_time=data.get("StateChangeTime"),
        region=data.get("Region", os.environ.get("AWS_REGION", "us-east-1")),
        metric_name=trigger.get("MetricName"),
        namespace=trigger.get("Namespace"),
        dimensions=dimensions,
        statistic=trigger.get("Statistic"),
        period=trigger.get("Period"),
        threshold=trigger.get("Threshold"),
        comparison_operator=trigger.get("ComparisonOperator"),
    )
